File: php-api/src/Services/StrategyMetricService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Support\Uuid;
use PDO;

/**
 * Ported from Services/StrategyMetricService.cs. Metric CRUD (enforcing the exactly-one-parent rule:
 * exactly one of PillarId/EnablerId non-null, app-layer only per this codebase's no-CHECK-constraints
 * convention) plus append-only StrategyMetricEntry recording/history.
 */
final class StrategyMetricService
{
    public function __construct(private readonly PDO $db)
    {
    }

    private function resolvePillarOrg(string $pillarId): ?string
    {
        $stmt = $this->db->prepare(<<<SQL
            SELECT s."OrganisationId" FROM "StrategyPillars" p
            JOIN "Strategies" s ON s."Id" = p."StrategyId"
            WHERE p."Id" = :id
        SQL);
        $stmt->execute(['id' => $pillarId]);
        $org = $stmt->fetchColumn();
        return $org !== false ? $org : null;
    }

    private function resolveEnablerOrg(string $enablerId): ?string
    {
        $stmt = $this->db->prepare(<<<SQL
            SELECT s."OrganisationId" FROM "StrategyEnablers" e
            JOIN "StrategyPillars" p ON p."Id" = e."PillarId"
            JOIN "Strategies" s ON s."Id" = p."StrategyId"
            WHERE e."Id" = :id
        SQL);
        $stmt->execute(['id' => $enablerId]);
        $org = $stmt->fetchColumn();
        return $org !== false ? $org : null;
    }

    private function findOwnedMetric(string $organisationId, string $metricId): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "StrategyMetrics" WHERE "Id" = :id');
        $stmt->execute(['id' => $metricId]);
        $metric = $stmt->fetch();
        if ($metric === false) {
            return null;
        }

        $owningOrg = $metric['PillarId'] !== null
            ? $this->resolvePillarOrg($metric['PillarId'])
            : $this->resolveEnablerOrg($metric['EnablerId']);

        return $owningOrg === $organisationId ? $metric : null;
    }

    public function create(string $organisationId, ?string $pillarId, ?string $enablerId, array $request): ?array
    {
        // Exactly one parent — never both, never neither.
        if (($pillarId === null) === ($enablerId === null)) {
            return null;
        }

        if ($pillarId !== null) {
            if ($this->resolvePillarOrg($pillarId) !== $organisationId) {
                return null;
            }
        } else {
            if ($this->resolveEnablerOrg($enablerId) !== $organisationId) {
                return null;
            }
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $unitLabel = trim((string) ($request['unitLabel'] ?? '')) !== '' ? trim((string) $request['unitLabel']) : null;
        if ($unitLabel !== null && strlen($unitLabel) > 20) {
            $unitLabel = substr($unitLabel, 0, 20);
        }
        $targetValue = isset($request['targetValue']) && $request['targetValue'] !== null ? (float) $request['targetValue'] : null;

        $maxStmt = $this->db->prepare('SELECT MAX("SortOrder") FROM "StrategyMetrics" WHERE ' . ($pillarId !== null ? '"PillarId" = :parentId' : '"EnablerId" = :parentId'));
        $maxStmt->execute(['parentId' => $pillarId ?? $enablerId]);
        $max = $maxStmt->fetchColumn();
        $sortOrder = ($max !== false && $max !== null) ? ((int) $max) + 1 : 0;

        $metricId = Uuid::v4();
        $this->db->prepare('INSERT INTO "StrategyMetrics" ("Id", "PillarId", "EnablerId", "Name", "TargetValue", "UnitLabel", "SortOrder") VALUES (:id, :pillarId, :enablerId, :name, :targetValue, :unitLabel, :sortOrder)')
            ->execute(['id' => $metricId, 'pillarId' => $pillarId, 'enablerId' => $enablerId, 'name' => $name, 'targetValue' => $targetValue, 'unitLabel' => $unitLabel, 'sortOrder' => $sortOrder]);

        return ['id' => $metricId, 'pillarId' => $pillarId, 'enablerId' => $enablerId, 'name' => $name, 'targetValue' => $targetValue, 'unitLabel' => $unitLabel, 'sortOrder' => $sortOrder];
    }

    public function update(string $organisationId, string $metricId, array $request): ?array
    {
        $metric = $this->findOwnedMetric($organisationId, $metricId);
        if ($metric === null) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            return null;
        }
        if (strlen($name) > 150) {
            $name = substr($name, 0, 150);
        }
        $unitLabel = trim((string) ($request['unitLabel'] ?? '')) !== '' ? trim((string) $request['unitLabel']) : null;
        if ($unitLabel !== null && strlen($unitLabel) > 20) {
            $unitLabel = substr($unitLabel, 0, 20);
        }
        $targetValue = isset($request['targetValue']) && $request['targetValue'] !== null ? (float) $request['targetValue'] : null;
        $sortOrder = (int) ($request['sortOrder'] ?? $metric['SortOrder']);

        $this->db->prepare('UPDATE "StrategyMetrics" SET "Name" = :name, "TargetValue" = :targetValue, "UnitLabel" = :unitLabel, "SortOrder" = :sortOrder WHERE "Id" = :id')
            ->execute(['name' => $name, 'targetValue' => $targetValue, 'unitLabel' => $unitLabel, 'sortOrder' => $sortOrder, 'id' => $metricId]);

        return ['id' => $metricId, 'pillarId' => $metric['PillarId'], 'enablerId' => $metric['EnablerId'], 'name' => $name, 'targetValue' => $targetValue, 'unitLabel' => $unitLabel, 'sortOrder' => $sortOrder];
    }

    public function delete(string $organisationId, string $metricId): bool
    {
        if ($this->findOwnedMetric($organisationId, $metricId) === null) {
            return false;
        }
        $this->db->prepare('DELETE FROM "StrategyMetrics" WHERE "Id" = :id')->execute(['id' => $metricId]);
        return true;
    }

    // ---- Metric entries (append-only time series) ----

    public function recordEntry(string $organisationId, string $metricId, array $request): ?array
    {
        if ($this->findOwnedMetric($organisationId, $metricId) === null) {
            return null;
        }

        $entryId = Uuid::v4();
        $value = (float) ($request['value'] ?? 0);
        $note = trim((string) ($request['note'] ?? '')) !== '' ? trim((string) $request['note']) : null;

        $this->db->prepare('INSERT INTO "StrategyMetricEntries" ("Id", "MetricId", "RecordedAt", "Value", "Note") VALUES (:id, :metricId, now(), :value, :note)')
            ->execute(['id' => $entryId, 'metricId' => $metricId, 'value' => $value, 'note' => $note]);

        $stmt = $this->db->prepare('SELECT "RecordedAt" FROM "StrategyMetricEntries" WHERE "Id" = :id');
        $stmt->execute(['id' => $entryId]);
        $recordedAt = $stmt->fetchColumn();

        return ['id' => $entryId, 'metricId' => $metricId, 'recordedAt' => $recordedAt, 'value' => $value, 'note' => $note];
    }

    public function getHistory(string $organisationId, string $metricId): ?array
    {
        if ($this->findOwnedMetric($organisationId, $metricId) === null) {
            return null;
        }
        return $this->fetchHistory($metricId);
    }

    /** ProjectMember-readable variant — resolves the project's own org first (the caller has no
     * organisationId of their own, only a projectId ProjectMemberMiddleware already verified they
     * belong to), then re-uses the same ownership-checked getHistory. */
    public function getHistoryForProject(string $projectId, string $metricId): ?array
    {
        $stmt = $this->db->prepare('SELECT "OrganisationId" FROM "Projects" WHERE "Id" = :id');
        $stmt->execute(['id' => $projectId]);
        $organisationId = $stmt->fetchColumn();
        if ($organisationId === false) {
            return null;
        }
        return $this->getHistory($organisationId, $metricId);
    }

    private function fetchHistory(string $metricId): array
    {
        $stmt = $this->db->prepare('SELECT "Id", "MetricId", "RecordedAt", "Value", "Note" FROM "StrategyMetricEntries" WHERE "MetricId" = :id ORDER BY "RecordedAt"');
        $stmt->execute(['id' => $metricId]);
        return array_map(static fn(array $e): array => [
            'id' => $e['Id'], 'metricId' => $e['MetricId'], 'recordedAt' => $e['RecordedAt'],
            'value' => (float) $e['Value'], 'note' => $e['Note'],
        ], $stmt->fetchAll());
    }
}

File: api/Enkl.Api/Services/StrategyMetricService.cs
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

/// <summary>
/// Metric CRUD (enforcing the exactly-one-parent rule: exactly one of PillarId/EnablerId non-null,
/// app-layer only per root CLAUDE.md's no-CHECK-constraints convention) plus append-only
/// StrategyMetricEntry recording/history. Ownership re-validation always resolves back to
/// OrganisationId regardless of which parent the metric has.
/// </summary>
public class StrategyMetricService
{
    private readonly AppDbContext _db;

    public StrategyMetricService(AppDbContext db)
    {
        _db = db;
    }

    private async Task<Guid?> ResolvePillarOrgAsync(Guid pillarId) =>
        await _db.StrategyPillars.Where(p => p.Id == pillarId).Select(p => (Guid?)p.Strategy.OrganisationId).FirstOrDefaultAsync();

    private async Task<Guid?> ResolveEnablerOrgAsync(Guid enablerId) =>
        await _db.StrategyEnablers.Where(e => e.Id == enablerId).Select(e => (Guid?)e.Pillar.Strategy.OrganisationId).FirstOrDefaultAsync();

    private async Task<StrategyMetric?> GetOwnedMetricAsync(Guid organisationId, Guid metricId)
    {
        var metric = await _db.StrategyMetrics
            .Include(m => m.Pillar!).ThenInclude(p => p.Strategy)
            .Include(m => m.Enabler!).ThenInclude(e => e.Pillar).ThenInclude(p => p.Strategy)
            .FirstOrDefaultAsync(m => m.Id == metricId);
        if (metric is null) return null;

        var owningOrg = metric.PillarId != null ? metric.Pillar!.Strategy.OrganisationId : metric.Enabler!.Pillar.Strategy.OrganisationId;
        return owningOrg == organisationId ? metric : null;
    }

    public async Task<StrategyMetricDto?> CreateAsync(Guid organisationId, Guid? pillarId, Guid? enablerId, CreateStrategyMetricRequest request)
    {
        // Exactly one parent — never both, never neither.
        if ((pillarId is null) == (enablerId is null)) return null;

        if (pillarId is not null)
        {
            var org = await ResolvePillarOrgAsync(pillarId.Value);
            if (org != organisationId) return null;
        }
        else
        {
            var org = await ResolveEnablerOrgAsync(enablerId!.Value);
            if (org != organisationId) return null;
        }

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        var unitLabel = string.IsNullOrWhiteSpace(request.UnitLabel) ? null : request.UnitLabel!.Trim();
        if (unitLabel != null && unitLabel.Length > 20) unitLabel = unitLabel[..20];

        var maxSortOrder = await _db.StrategyMetrics
            .Where(m => pillarId != null ? m.PillarId == pillarId : m.EnablerId == enablerId)
            .Select(m => (int?)m.SortOrder).MaxAsync() ?? -1;

        var metric = new StrategyMetric
        {
            Id = Guid.NewGuid(),
            PillarId = pillarId,
            EnablerId = enablerId,
            Name = name,
            TargetValue = request.TargetValue,
            UnitLabel = unitLabel,
            SortOrder = maxSortOrder + 1
        };
        _db.StrategyMetrics.Add(metric);
        await _db.SaveChangesAsync();
        return ToDto(metric);
    }

    public async Task<StrategyMetricDto?> UpdateAsync(Guid organisationId, Guid metricId, UpdateStrategyMetricRequest request)
    {
        var metric = await GetOwnedMetricAsync(organisationId, metricId);
        if (metric is null) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) return null;
        if (name.Length > 150) name = name[..150];

        var unitLabel = string.IsNullOrWhiteSpace(request.UnitLabel) ? null : request.UnitLabel!.Trim();
        if (unitLabel != null && unitLabel.Length > 20) unitLabel = unitLabel[..20];

        metric.Name = name;
        metric.TargetValue = request.TargetValue;
        metric.UnitLabel = unitLabel;
        metric.SortOrder = request.SortOrder;
        await _db.SaveChangesAsync();
        return ToDto(metric);
    }

    public async Task<bool> DeleteAsync(Guid organisationId, Guid metricId)
    {
        var metric = await GetOwnedMetricAsync(organisationId, metricId);
        if (metric is null) return false;

        _db.StrategyMetrics.Remove(metric);
        await _db.SaveChangesAsync();
        return true;
    }

    // ---- Metric entries (append-only time series) ----

    public async Task<StrategyMetricEntryDto?> RecordEntryAsync(Guid organisationId, Guid metricId, CreateStrategyMetricEntryRequest request)
    {
        var metric = await GetOwnedMetricAsync(organisationId, metricId);
        if (metric is null) return null;

        var entry = new StrategyMetricEntry
        {
            Id = Guid.NewGuid(),
            MetricId = metricId,
            RecordedAt = DateTime.UtcNow,
            Value = request.Value,
            Note = string.IsNullOrWhiteSpace(request.Note) ? null : request.Note!.Trim()
        };
        _db.StrategyMetricEntries.Add(entry);
        await _db.SaveChangesAsync();
        return new StrategyMetricEntryDto(entry.Id, entry.MetricId, entry.RecordedAt, entry.Value, entry.Note);
    }

    public async Task<List<StrategyMetricEntryDto>?> GetHistoryAsync(Guid organisationId, Guid metricId)
    {
        var metric = await GetOwnedMetricAsync(organisationId, metricId);
        if (metric is null) return null;

        return await _db.StrategyMetricEntries.AsNoTracking()
            .Where(e => e.MetricId == metricId)
            .OrderBy(e => e.RecordedAt)
            .Select(e => new StrategyMetricEntryDto(e.Id, e.MetricId, e.RecordedAt, e.Value, e.Note))
            .ToListAsync();
    }

    /// <summary>ProjectMember-readable variant — resolves the project's own org first (the caller has
    /// no organisationId of their own, only a projectId ProjectMemberAuthorizationHandler already
    /// verified they belong to), then re-uses the same ownership-checked GetHistoryAsync so a metric
    /// from a different org can't be probed via this route either.</summary>
    public async Task<List<StrategyMetricEntryDto>?> GetHistoryForProjectAsync(Guid projectId, Guid metricId)
    {
        var organisationId = await _db.Projects.Where(p => p.Id == projectId).Select(p => (Guid?)p.OrganisationId).FirstOrDefaultAsync();
        if (organisationId is null) return null;
        return await GetHistoryAsync(organisationId.Value, metricId);
    }

    private static StrategyMetricDto ToDto(StrategyMetric m) => new(m.Id, m.PillarId, m.EnablerId, m.Name, m.TargetValue, m.UnitLabel, m.SortOrder);
}
